[NCCL] Init dummy NCCL comms in constructor

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #65173

Initializes dummy NCCL communicators in constructor for a basic health
check that communicators can be initialized prior to launching the first
collective.

After successful init, we immediately use `ncclCommAbort` to destroy these
communicators to ensure they don't interfere with regular communicator creation
during collectives.

For failures, the reason is mostly hangs when initializing the NCCL comm. To detect this it is done in a separate thread, and the main thread does a timed wait for a successful return and throws if it encounters a timeout. Other errors will just terminate the thread resulting in an application crash. 

Differential Revision: [D31005792](https://our.internmc.facebook.com/intern/diff/D31005792/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31005792/)!

rohan-varma committed:
[NCCL] Init dummy NCCL comms in constructor

Initializes dummy NCCL communicators in constructor for a basic health
check that communicators can be initialized prior to launching the first
collective.

After successful init, we immediately use `ncclCommAbort` to destroy these
communicators to ensure they don't interfere with regular communicator creation
during collectives.

Differential Revision: [D31005792](https://our.internmc.facebook.com/intern/diff/D31005792/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31005792/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[NCCL] Init dummy NCCL comms in constructor"


Initializes dummy NCCL communicators in constructor for a basic health
check that communicators can be initialized prior to launching the first
collective.

After successful init, we immediately use `ncclCommAbort` to destroy these
communicators to ensure they don't interfere with regular communicator creation
during collectives.

Differential Revision: [D31005792](https://our.internmc.facebook.com/intern/diff/D31005792/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31005792/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[NCCL] Init dummy NCCL comms in constructor"


Initializes dummy NCCL communicators in constructor for a basic health
check that communicators can be initialized prior to launching the first
collective.

After successful init, we immediately use `ncclCommAbort` to destroy these
communicators to ensure they don't interfere with regular communicator creation
during collectives.

For failures, the reason is mostly hangs when initializing the NCCL comm. To detect this it is done in a separate thread, and the main thread does a timed wait for a successful return and throws if it encounters a timeout. Other errors will just terminate the thread resulting in an application crash. 

Differential Revision: [D31005792](https://our.internmc.facebook.com/intern/diff/D31005792/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31005792/)!

[ghstack-poisoned]

pritamdamania87 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -482,6 +488,43 @@ ProcessGroupNCCL::ProcessGroupNCCL(
     asyncErrorHandling_ = false;
   }
 
+  // Perform health check by initializing dummy communicators and destroying
```
nit: move this logic to a separate helper method.

pritamdamania87 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -482,6 +488,43 @@ ProcessGroupNCCL::ProcessGroupNCCL(
+  // collective.
+  // Run it in a separate thread and wait on CV to handle timeouts, since
+  // majority of getNCCLComm failures are hangs.
+  std::mutex healthCheckMutex;
```
nit: can use atomic bool for healthCheckSuccess and no need for this mutex then?

rohan-varma replied:
I think we still need the lock due to the condition variable requiring a lock? 

Also I read this article: https://www.modernescpp.com/index.php/c-core-guidelines-be-aware-of-the-traps-of-condition-variables that said an atomic var is not enough to avoid races in some case. Most of the logic around signaling the main thread comes from there, does it seem okay?

pritamdamania87 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -482,6 +488,43 @@ ProcessGroupNCCL::ProcessGroupNCCL(
+  // CV. Hence we detach the thread here.
+  t.detach();
+  std::unique_lock<std::mutex> lock(healthCheckMutex);
+  LOG(INFO) << "Rank: " << rank_ << " will wait for: " << options->timeout.count() << " ms for NCCL health check to complete.";
```
nit: This format is slightly different from the format followed in all other places: https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp#L295

pritamdamania87 replied:
Probably not a good idea to perform logging under a lock.

rohan-varma replied:
Good catch, will remove it to before the lock.

pritamdamania87 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -482,6 +488,43 @@ ProcessGroupNCCL::ProcessGroupNCCL(
+  std::condition_variable healthCheckCv;
+  bool healthCheckSuccess = false;
+  auto t = std::thread([&healthCheckMutex, &healthCheckCv, &healthCheckSuccess, this]() {
+    std::vector<at::Device> rankDevice = {getDeviceForRank(rank_)};
```
We probably need some signaling mechanism from this thread to the main thread. If this thread dies due to an exception, the main thread should know about it and it should be reported. Probably need to catch the exception and store it in an exception_ptr that can be queried by the main thread.

rohan-varma replied:
I mentioned this on 494-495, but the vast majority of errors we expect here (and seen in internal training workloads) is hangs. As a result we don't get an exception we can signal the main thread with. 

To deal with the hang, I added a wait on `healthCheckCv` so that if this thread doesn't signal in time (whether it died due to exception or hang), we get a timeout on the main thread and indicate the health check failed. 

Seems like this is enough to indicate something went wrong with communicator initialization, but in the cases of exceptions I guess we could wrap all this in a try-catch and populate an `exception_ptr`?

pritamdamania87 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -482,6 +488,43 @@ ProcessGroupNCCL::ProcessGroupNCCL(
+      lock, options->timeout, [&healthCheckSuccess]() { return healthCheckSuccess; });
+
+  if (!healthCheckSuccess) {
+    LOG(INFO) << "Detected health check failure.";
```
This seems redundant given `TORCH_CHECK` below. Also, isn't it simpler to just do this:

```
TORCH_CHECK(healthCheckSuccess, "ProcessGroupNCCL: Health check falure: Failed to initialize NCCL communicator on rank ", rank_);
```

pritamdamania87 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -854,6 +897,27 @@ void ProcessGroupNCCL::broadcastUniqueNCCLID(
 
+
+void ProcessGroupNCCL::destroyNCCLComms(const std::string& devNCCLCommMapKey) {
+  if (devNCCLCommMap_.find(devNCCLCommMapKey) == devNCCLCommMap_.end()) {
```
Don't you need a lock here to access `devNCCLCommMap_`?

pritamdamania87 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -854,6 +897,27 @@ void ProcessGroupNCCL::broadcastUniqueNCCLID(
+      devNCCLCommMap_[devNCCLCommMapKey];
+  // Loop through communicators and call ncclCommAbort.
+  for (const auto& comm : ncclComms) {
+    // ncclCommDestroy(comm->getNcclComm()) results in segfault when PG is being
```
Could you elaborate a bit more on why this happens?

rohan-varma replied:
Some segfault issues are mentioned on L43 here: https://caffe2.ai/doxygen-c/html/nccl_8cpp_source.html, as well as in PT - https://github.com/pytorch/pytorch/blob/master/torch/csrc/cuda/nccl.cpp#L373. 

Although, it is a bit strange since this isn't called in dtor but still getting a segfault (in `~ProcessGroupNCCL`) when enabling this, and calling `ncclCommAbort` seems to have fixed the issue.

pritamdamania87 commented on test/cpp/c10d/ProcessGroupNCCLTest.cpp:
```diff
@@ -497,6 +503,36 @@ void testReduceScatter(const std::string& path, int rank, int size) {
+    // Catch error relating to health check failure
+    bool error_caught = false;
+    try {
+      test.initialize(0, worldSize);
```
So the error occurs since every rank specifies rank 0?

rohan-varma replied:
Yep, this is to simulate a hang which the main thread should eventually become aware of (after the 3s timeout).

pritamdamania87 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -139,6 +140,11 @@ std::vector<at::Device> getDeviceList(const std::vector<at::Tensor>& tensors) {
 
+// Return CUDA device with ordinal given by input rank.
+at::Device getDeviceForRank(int rank) {
+  return at::Device(at::DeviceType::CUDA, rank);
```
This probably breaks if rank is something like 13, don't you need rank % numGPUs similar to what barrier does?

rohan-varma replied:
Yes, thanks for catching that!

rohan-varma committed:
Update on "[NCCL] Init dummy NCCL comms in constructor"


Initializes dummy NCCL communicators in constructor for a basic health
check that communicators can be initialized prior to launching the first
collective.

After successful init, we immediately use `ncclCommAbort` to destroy these
communicators to ensure they don't interfere with regular communicator creation
during collectives.

For failures, the reason is mostly hangs when initializing the NCCL comm. To detect this it is done in a separate thread, and the main thread does a timed wait for a successful return and throws if it encounters a timeout. Other errors will just terminate the thread resulting in an application crash. 

Differential Revision: [D31005792](https://our.internmc.facebook.com/intern/diff/D31005792/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31005792/)!

[ghstack-poisoned]

rohan-varma commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -139,6 +141,14 @@ std::vector<at::Device> getDeviceList(const std::vector<at::Tensor>& tensors) {
 
+// Return CUDA device with ordinal given by input rank.
+at::Device getDeviceForRank(int rank) {
+  TORCH_CHECK(rank >= 0, "Invalid rank ", rank);
```
This is mostly here for the test - to simulate the case where an exception (rather than timeout) happens in health check.

rohan-varma committed:
Update on "[NCCL] Init dummy NCCL comms in constructor"


Initializes dummy NCCL communicators in constructor for a basic health
check that communicators can be initialized prior to launching the first
collective.

After successful init, we immediately use `ncclCommAbort` to destroy these
communicators to ensure they don't interfere with regular communicator creation
during collectives.

For failures, the reason is mostly hangs when initializing the NCCL comm. To detect this it is done in a separate thread, and the main thread does a timed wait for a successful return and throws if it encounters a timeout. Other errors will just terminate the thread resulting in an application crash. 

Differential Revision: [D31005792](https://our.internmc.facebook.com/intern/diff/D31005792/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31005792/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[NCCL] Init dummy NCCL comms in constructor"


Initializes dummy NCCL communicators in constructor for a basic health
check that communicators can be initialized prior to launching the first
collective.

After successful init, we immediately use `ncclCommAbort` to destroy these
communicators to ensure they don't interfere with regular communicator creation
during collectives.

For failures, the reason is mostly hangs when initializing the NCCL comm. To detect this it is done in a separate thread, and the main thread does a timed wait for a successful return and throws if it encounters a timeout. Other errors will just terminate the thread resulting in an application crash. 

Differential Revision: [D31005792](https://our.internmc.facebook.com/intern/diff/D31005792/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31005792/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[NCCL] Init dummy NCCL comms in constructor"


Initializes dummy NCCL communicators in constructor for a basic health
check that communicators can be initialized prior to launching the first
collective.

After successful init, we immediately use `ncclCommAbort` to destroy these
communicators to ensure they don't interfere with regular communicator creation
during collectives.

For failures, the reason is mostly hangs when initializing the NCCL comm. To detect this it is done in a separate thread, and the main thread does a timed wait for a successful return and throws if it encounters a timeout. Other errors will just terminate the thread resulting in an application crash. 

Differential Revision: [D31005792](https://our.internmc.facebook.com/intern/diff/D31005792/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31005792/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[NCCL] Init dummy NCCL comms in constructor"


Initializes dummy NCCL communicators in constructor for a basic health
check that communicators can be initialized prior to launching the first
collective.

After successful init, we immediately use `ncclCommAbort` to destroy these
communicators to ensure they don't interfere with regular communicator creation
during collectives.

For failures, the reason is mostly hangs when initializing the NCCL comm. To detect this it is done in a separate thread, and the main thread does a timed wait for a successful return and throws if it encounters a timeout. Other errors will just terminate the thread resulting in an application crash. 

Differential Revision: [D31005792](https://our.internmc.facebook.com/intern/diff/D31005792/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31005792/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[NCCL] Init dummy NCCL comms in constructor"


Initializes dummy NCCL communicators in constructor for a basic health
check that communicators can be initialized prior to launching the first
collective.

After successful init, we immediately use `ncclCommAbort` to destroy these
communicators to ensure they don't interfere with regular communicator creation
during collectives.

For failures, the reason is mostly hangs when initializing the NCCL comm. To detect this it is done in a separate thread, and the main thread does a timed wait for a successful return and throws if it encounters a timeout. Other errors will just terminate the thread resulting in an application crash. 

Differential Revision: [D31005792](https://our.internmc.facebook.com/intern/diff/D31005792/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31005792/)!

[ghstack-poisoned]

pritamdamania87 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -527,6 +544,62 @@ ProcessGroupNCCL::ProcessGroupNCCL(
+      // Populate exception ptr and boolean indicating it is set.
+      // TODO: there does not appear to be a way to propagate the type of e like
+      // python, so just using runtime_error here.
+      healthCheckData.healthCheckException = std::make_exception_ptr(std::runtime_error(c10::str("Health check got exception while initializing + destroying communicators: ", e.what())));
```
Why not just do `healthCheckException = std::current_exception`?

pritamdamania87 reviewed (approved):
LGTM, although look at my previous comment regarding just using `std::current_exception` to capture the exception.

rohan-varma committed:
Update on "[NCCL] Init dummy NCCL comms in constructor"


Initializes dummy NCCL communicators in constructor for a basic health
check that communicators can be initialized prior to launching the first
collective.

After successful init, we immediately use `ncclCommAbort` to destroy these
communicators to ensure they don't interfere with regular communicator creation
during collectives.

For failures, the reason is mostly hangs when initializing the NCCL comm. To detect this it is done in a separate thread, and the main thread does a timed wait for a successful return and throws if it encounters a timeout. Other errors will just terminate the thread resulting in an application crash. 

Differential Revision: [D31005792](https://our.internmc.facebook.com/intern/diff/D31005792/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31005792/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[NCCL] Init dummy NCCL comms in constructor"


Initializes dummy NCCL communicators in constructor for a basic health
check that communicators can be initialized prior to launching the first
collective.

After successful init, we immediately use `ncclCommAbort` to destroy these
communicators to ensure they don't interfere with regular communicator creation
during collectives.

For failures, the reason is mostly hangs when initializing the NCCL comm. To detect this it is done in a separate thread, and the main thread does a timed wait for a successful return and throws if it encounters a timeout. Other errors will just terminate the thread resulting in an application crash. 

Differential Revision: [D31005792](https://our.internmc.facebook.com/intern/diff/D31005792/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31005792/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[NCCL] Init dummy NCCL comms in constructor"


Initializes dummy NCCL communicators in constructor for a basic health
check that communicators can be initialized prior to launching the first
collective.

After successful init, we immediately use `ncclCommAbort` to destroy these
communicators to ensure they don't interfere with regular communicator creation
during collectives.

For failures, the reason is mostly hangs when initializing the NCCL comm. To detect this it is done in a separate thread, and the main thread does a timed wait for a successful return and throws if it encounters a timeout. Other errors will just terminate the thread resulting in an application crash. 

Differential Revision: [D31005792](https://our.internmc.facebook.com/intern/diff/D31005792/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31005792/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[NCCL] Init dummy NCCL comms in constructor"


Initializes dummy NCCL communicators in constructor for a basic health
check that communicators can be initialized prior to launching the first
collective.

After successful init, we immediately use `ncclCommAbort` to destroy these
communicators to ensure they don't interfere with regular communicator creation
during collectives.

For failures, the reason is mostly hangs when initializing the NCCL comm. To detect this it is done in a separate thread, and the main thread does a timed wait for a successful return and throws if it encounters a timeout. Other errors will just terminate the thread resulting in an application crash. 

Differential Revision: [D31005792](https://our.internmc.facebook.com/intern/diff/D31005792/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31005792/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[NCCL] Init dummy NCCL comms in constructor"


Initializes dummy NCCL communicators in constructor for a basic health
check that communicators can be initialized prior to launching the first
collective.

After successful init, we immediately use `ncclCommAbort` to destroy these
communicators to ensure they don't interfere with regular communicator creation
during collectives.

For failures, the reason is mostly hangs when initializing the NCCL comm. To detect this it is done in a separate thread, and the main thread does a timed wait for a successful return and throws if it encounters a timeout. Other errors will just terminate the thread resulting in an application crash. 

Differential Revision: [D31005792](https://our.internmc.facebook.com/intern/diff/D31005792/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31005792/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[NCCL] Init dummy NCCL comms in constructor"


Initializes dummy NCCL communicators in constructor for a basic health
check that communicators can be initialized prior to launching the first
collective.

After successful init, we immediately use `ncclCommAbort` to destroy these
communicators to ensure they don't interfere with regular communicator creation
during collectives.

For failures, the reason is mostly hangs when initializing the NCCL comm. To detect this it is done in a separate thread, and the main thread does a timed wait for a successful return and throws if it encounters a timeout. Other errors will just terminate the thread resulting in an application crash. 

Differential Revision: [D31005792](https://our.internmc.facebook.com/intern/diff/D31005792/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31005792/)!

[ghstack-poisoned]

mruberry commented:
Unlanding. This appears to be breaking _linux-xenial-cuda10.2-py3.6-gcc7 / test (multigpu, 1, 1, linux.16xlarge.nvidia.gpu). See https://github.com/pytorch/pytorch/runs/3750225292. Relevant snippet:

```
 ERROR: CommTest.test_nccl_barrier_timeout

Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 423, in wrapper
    self._join_processes(fn)
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 642, in _join_processes
    self._check_return_codes(elapsed_time)
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 694, in _check_return_codes
    i, elapsed_time
RuntimeError: Process 0 terminated or timed out after 100.02010369300842 seconds
		
🚨 ERROR: CommTest.test_nccl_barrier_timeout_new_group

Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 423, in wrapper
    self._join_processes(fn)
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 642, in _join_processes
    self._check_return_codes(elapsed_time)
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 687, in _check_return_codes
    raise RuntimeError(error)
RuntimeError: Process 1 exited with error code 10 and exception:
Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 655, in _check_no_test_errors
    "Process {} timed out after {} seconds".format(i, elapsed_time)
RuntimeError: Process 0 timed out after 2.506455659866333 seconds
		
🚨 ERROR: NcclErrorHandlingTest.test_nccl_timeout

Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 423, in wrapper
    self._join_processes(fn)
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 642, in _join_processes
    self._check_return_codes(elapsed_time)
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 687, in _check_return_codes
    raise RuntimeError(error)
RuntimeError: Process 1 exited with error code 10 and exception:
Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 541, in run_test
    getattr(self, test_name)()
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 425, in wrapper
    fn()
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 175, in wrapper
    ret = func(*args, **kwargs)
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 2848, in wrapper
    return func(*args, **kwargs)
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 112, in wrapper
    return func(*args, **kwargs)
  File "/var/lib/jenkins/workspace/test/distributed/test_c10d_nccl.py", line 2389, in test_nccl_timeout
    store, self.rank, self.world_size, timeout=timedelta(seconds=timeout)
RuntimeError: ProcessGroupNCCL: Health check falure: Failed to initialize NCCL communicator on rank 1

Process 2 exited with error code 10 and exception:
Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 541, in run_test
    getattr(self, test_name)()
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 425, in wrapper
    fn()
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 175, in wrapper
    ret = func(*args, **kwargs)
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 2848, in wrapper
    return func(*args, **kwargs)
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 112, in wrapper
    return func(*args, **kwargs)
  File "/var/lib/jenkins/workspace/test/distributed/test_c10d_nccl.py", line 2389, in test_nccl_timeout
    store, self.rank, self.world_size, timeout=timedelta(seconds=timeout)
RuntimeError: ProcessGroupNCCL: Health check falure: Failed to initialize NCCL communicator on rank 2
```

@rohan-varma maybe we need to improve how/when that job is triggered?

rohan-varma commented:
resubmit in https://github.com/pytorch/pytorch/pull/65883
